lib/ruby_lsp/requests/document_highlight.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # The [document highlight](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight)
+    # informs the editor all relevant elements of the currently pointed item for highlighting. For example, when
+    # the cursor is on the `F` of the constant `FOO`, the editor should identify other occurences of `FOO`
+    # and highlight them.
+    #
+    # For writable elements like constants or variables, their read/write occurrences should be highlighted differently.
+    # This is achieved by sending different "kind" attributes to the editor (2 for read and 3 for write).
+    #
+    # # Example
+    #
+    # ```ruby
+    # FOO = 1 # should be highlighted as "write"
+    #
+    # def foo
+    #   FOO # should be highlighted as "read"
+    # end
+    # ````
+    class DocumentHighlight < BaseRequest
+      def self.run(document, position)
+        new(document, position).run
+      end
+
+      def initialize(document, position)
+        @highlights = []
+        position = Document::Scanner.new(document.source).find_position(position)
+        @target = find(document.tree, position)
+
+        super(document)
+      end
+
+      def run
+        # no @target means the target is not highlightable
+        return [] unless @target
+
+        visit(@document.tree)
+        @highlights
+      end
+
+      def visit_var_field(node)
+        add_highlight(
+          node.value,
+          LanguageServer::Protocol::Constant::DocumentHighlightKind::WRITE
+        ) if matches_target?(node.value)
```

Comment:
This is super nitpicky, but can we flip this around, since I think post conditionals are really hard to read on multiline statements?
```suggestion
        return unless matches_target?(node.value)
        
        add_highlight(
          node.value,
          LanguageServer::Protocol::Constant::DocumentHighlightKind::WRITE
        )
```

Reply:
Also, should it be `matches_target?(node)` or `matches_target?(node.value)`?

Reply:
It's `matches_target?(node.value)` because the node here is a `VarField` or `VarRef`. And their value (e.g. `Ident`) is what we try to match.